### ISSUE TITLE
tests: flaky deadlock in switch TestChannelLinkCancelFullCommitment test

### ISSUE THREAD

**Issue by Roasbeef** [lightningnetwork/lnd#5437]
There seems to be some spooky concurrency interaction happenign within the switch that's either unique to the way the test context in `TestChannelLinkCancelFullCommitment` is set up, or inherent w/ the current [condition variable usage within the mailbox](https://github.com/lightningnetwork/lnd/blob/master/htlcswitch/mailbox.go#L358). 

A sample stack trace when the test times out is: 
```
goroutine 104 [select, 7 minutes]:
github.com/lightningnetwork/lnd/htlcswitch.(*memoryMailBox).mailCourier(0xc0001fc4d0, 0xc00024f700)
	/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/htlcswitch/mailbox.go:453 +0x68d
created by github.com/lightningnetwork/lnd/htlcswitch.(*memoryMailBox).Start.func1
	/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/htlcswitch/mailbox.go:181 +0x48

goroutine 105 [sync.Cond.Wait, 9 minutes]:
sync.runtime_notifyListWait(0xc000564a10, 0x1)
	/Users/roasbeef/go/src/runtime/sema.go:513 +0xf8
sync.(*Cond).Wait(0xc000564a00)
	/Users/roasbeef/go/src/sync/cond.go:56 +0x99
github.com/lightningnetwork/lnd/htlcswitch.(*memoryMailBox).mailCourier(0xc0001fc4d0, 0xc00024f701)
	/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/htlcswitch/mailbox.go:382 +0xaa5
created by github.com/lightningnetwork/lnd/htlcswitch.(*memoryMailBox).Start.func1
	/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/htlcswitch/mailbox.go:182 +0x6f


goroutine 1580 [select, 8 minutes]:
github.com/lightningnetwork/lnd/htlcswitch.(*Switch).GetPaymentResult.func1(0xc00021cfc0, 0xc002816720, 0xc002816ea0, 0x832f88d4d835fddf, 0x16bf200, 0x1947498, 0xb1acd6f0b999b9d, 0x55e39d69e2443c9d, 0x73f532b3c4a78aba, 0xcc0b72d5a4cc7604)
	/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/htlcswitch/switch.go:416 +0xe5
created by github.com/lightningnetwork/lnd/htlcswitch.(*Switch).GetPaymentResult
	/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/htlcswitch/switch.go:412 +0x187

goroutine 1590 [select, 8 minutes]:
github.com/lightningnetwork/lnd/htlcswitch.(*Switch).GetPaymentResult.func1(0xc00021cfc0, 0xc00037a300, 0xc00037a960, 0x792f8f3e28150f88, 0x16bf200, 0x1947498, 0x50bca2d963cf10fe, 0xcdab8a4697708128, 0x799bb7e84141827c, 0x3cd4217204bc148f)
	/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/htlcswitch/switch.go:416 +0xe5
created by github.com/lightningnetwork/lnd/htlcswitch.(*Switch).GetPaymentResult
	/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/htlcswitch/switch.go:412 +0x187

goroutine 1607 [select, 8 minutes]:
github.com/lightningnetwork/lnd/htlcswitch.(*Switch).GetPaymentResult.func1(0xc00021cfc0, 0xc00279c480, 0xc00279c4e0, 0xb82f4d7196c12ed9, 0x16bf200, 0x1947498, 0xd4110826a9739be3, 0x5d18abe0d6715c44, 0x452208ecd5b2b6da, 0xed00e2c7cb9a15d9)
	/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/htlcswitch/switch.go:416 +0xe5
created by github.com/lightningnetwork/lnd/htlcswitch.(*Switch).GetPaymentResult
	/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/htlcswitch/switch.go:412 +0x187

goroutine 2027 [chan receive, 8 minutes]:
github.com/lightningnetwork/lnd/htlcswitch.(*twoHopNetwork).makeHoldPayment.func1(0xc0001546e0, 0xefad751f79159af5, 0xdadb6f7d7fc5512, 0x53a1c3e08949f2f0, 0xbd0e2d82e927fcf, 0xefb98d09219bde0e, 0xc000753020)
	/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/htlcswitch/test_utils.go:1343 +0xc5
created by github.com/lightningnetwork/lnd/htlcswitch.(*twoHopNetwork).makeHoldPayment
	/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/htlcswitch/test_utils.go:1334 +0x59e

goroutine 2029 [chan receive, 8 minutes]:
github.com/lightningnetwork/lnd/htlcswitch.(*twoHopNetwork).makeHoldPayment.func1(0xc0001546e0, 0xc5e5f374cb246c70, 0xae5a8472027da3c5, 0x2b9acc84fe27ec3a, 0x99e7677da27d9ebe, 0x6eedab54d8a921d4, 0xc000267f80)
	/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/htlcswitch/test_utils.go:1343 +0xc5
created by github.com/lightningnetwork/lnd/htlcswitch.(*twoHopNetwork).makeHoldPayment
	/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/htlcswitch/test_utils.go:1334 +0x59e

goroutine 2033 [chan receive, 8 minutes]:
github.com/lightningnetwork/lnd/htlcswitch.(*twoHopNetwork).makeHoldPayment.func1(0xc0001546e0, 0x778e85381016da1a, 0x6003c8d002591402, 0x9bf8551960b57b7e, 0x9e1fce7d2edf5ab8, 0x4ebe7bb3125a80f7, 0xc0016699e0)
	/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/htlcswitch/test_utils.go:1343 +0xc5
created by github.com/lightningnetwork/lnd/htlcswitch.(*twoHopNetwork).makeHoldPayment
	/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/htlcswitch/test_utils.go:1334 +0x59e
``` 

Here we see hundreds of goroutine stack up that should be settled back in series once we fill up the commitment transaction.

**Comment by Roasbeef** [comment]
Pasting some initial investigation by @carlaKC: 

I added a timeout to `TestChannelLinkCancelFullCommitment` (see [branch](https://github.com/carlaKC/lnd/tree/TestChannelLinkCancelFullCommitment-deadlock)) and took a mutex profile. I suspect that this may be a deadlock similair to [what we saw in mission control](https://github.com/lightningnetwork/lnd/pull/5116) a while back.

In this case:

- `AddInvoice` acquires `InvoiceRegistry` lock, then tries to acquire `db` lock
- `[x]` acquires `db` lock, then tries to acquire `InvoiceRegistry` loc

We're now deadlocked because `AddInvoice` can't ever get `db` block, and `[x]` can't ever get `InvoiceRegistry` lock. 

This is just a guess so far, because I can't find `[x]` yet. I'm not a profile-interpreting wizard, so if anything jumps out at you looking at this let me know. Otherwise will keep digging into it in my spare time.
